proof/refine/ARM/SchedContextInv_R.thy

```diff
@@ -1352,6 +1354,544 @@ lemma scheduleUsed_corres:
+     apply (clarsimp simp: vs_all_heap_simps)
+    apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+        apply (rule hoare_weaken_pre)
+         apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+          apply wpsimp
+         apply wpsimp
+        apply wpsimp
+       apply (rule hoare_weaken_pre)
+        apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+         apply wpsimp
+        apply wpsimp
+       apply wpsimp
+      apply (rule hoare_weaken_pre)
+       apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+        apply wpsimp
+       apply wpsimp
+      apply wpsimp
+     apply (rule hoare_weaken_pre)
+      apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+       apply (wpsimp wp: handle_overrun_loop_body_non_zero_refills)
+      apply wpsimp
+     apply clarsimp
+    apply (rule hoare_weaken_pre)
+     apply (rule_tac f=cur_sc in hoare_lift_Pf2)
+      apply (wpsimp wp: handle_overrun_loop_body_refills_unat_sum_equals_budget)
+     apply wpsimp
+    apply clarsimp
+   apply (clarsimp simp: pred_conj_def)
+   apply (intro hoare_vcg_conj_lift_pre_fix; (solves wpsimp)?)
+     apply (wpsimp wp: handleOverrunLoopBody_valid_objs')
+    apply (rule hoare_weaken_pre)
+     apply (rule_tac f=ksCurSc in hoare_lift_Pf)
+      apply wpsimp
+     apply wpsimp
+    apply wpsimp
+   apply (rule hoare_weaken_pre)
+    apply (rule_tac f=ksCurSc in hoare_lift_Pf)
+     apply wpsimp
+    apply wpsimp
+   apply wpsimp
+  apply (fastforce intro:  handle_overrun_loop_body_terminates)
+  done
+
+lemma handle_overrun_loop_is_active_sc[wp]:
+  "handle_overrun_loop usage \<lbrace>\<lambda>s. is_active_sc sc_ptr s\<rbrace>"
+  apply handle_overrun_loop_simple
+  done
+
+lemma get_refills_exs_valid[wp]:
+  "sc_at sc_ptr s \<Longrightarrow> \<lbrace>(=) s\<rbrace> get_refills sc_ptr \<exists>\<lbrace>\<lambda>r. (=) s\<rbrace>"
+  apply (clarsimp simp: get_refills_def)
+  apply (wpsimp wp: get_sched_context_exs_valid)
+   apply (erule sc_atD1)
+  apply simp
+  done
+
+(* FIXME RT: keep? move to Lib? *)
+lemma corres_when3:
+  "\<lbrakk>G = G'; G \<Longrightarrow> corres_underlying sr nf nf' dc P P' a c\<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' dc P P' (when G a) (when G' c)"
+  apply (frule corres_when2)
+   apply (fastforce simp: corres_underlying_def when_def)+
+  done
+
+lemma update_refill_hd_is_active_sc2[wp]:
+  "update_refill_hd sc_ptr f \<lbrace>is_active_sc2 sc_ptr'\<rbrace>"
+  apply (clarsimp simp: update_refill_hd_def)
+  apply (wpsimp wp: update_sched_context_wp)
+  apply (clarsimp simp: is_active_sc2_def obj_at_def opt_map_red)
+  done
+
+lemma refillBudgetCheck_corres:
+  "usage = usage'
+   \<Longrightarrow> corres dc ((\<lambda>s. sc_at (cur_sc s) s \<and> is_active_sc (cur_sc s) s
+                       \<and> valid_objs s
+                       \<and> pspace_aligned s \<and> pspace_distinct s)
+                  and (\<lambda>s. \<not> round_robin (cur_sc s) s \<and> valid_refills (cur_sc s) s))
+                 valid_objs'
```

Comment:
I am pretty sure this can be weakened to `valid_refills’`. Similarly for `handleOverrunLoop_corres` and so on. I did the same with `scheduleUsed_corres`, etc., in my `refillUnblockCheck_corres` PR.

Reply:
Did you want to merge the `refillUnblockCheck_corres` PR first and then I can make the changes to these new `corres` rules?

Reply:
Done